compiler/noirc_evaluator/src/ssa/opt/loop_invariant.rs

```diff
@@ -236,34 +253,68 @@ impl<'f> LoopInvariantContext<'f> {
-                    self.outer_induction_variables.get(&binary.rhs),
-                ) {
-                    (Some((lhs, _)), None, None, Some(upper_bound)) => (lhs, *upper_bound),
-                    (None, Some((rhs, _)), Some(upper_bound), None) => (*upper_bound, rhs),
-                    _ => return false,
+    /// Binary operations can take advantage of that our induction variable has a fixed maximum,
+    /// to be transformed from a checked operation to an unchecked operation.
+    /// Checked operations require more bytecode and thus we aim to minimize their usage wherever possible.
+    ///
+    /// If one side of a binary operation is a constant and the other is an induction variable
+    /// with a known upper bound, we know whether that binary operation will ever overflow.
+    /// If we determine that an overflow is not possible we can convert the checked operation to unchecked.
+    fn transform_to_unchecked_from_upper_bound(&mut self, instruction_id: InstructionId) {
+        if let Instruction::Binary(binary) = &self.inserter.function.dfg[instruction_id] {
+            let new_instruction = if !binary.operator.is_unchecked()
+                && self
+                    .eval_binary_op_with_induction_vars(binary, &self.current_induction_variables)
+            {
+                let binary = Binary {
+                    lhs: binary.lhs,
+                    rhs: binary.rhs,
+                    operator: binary.operator.into_unchecked(),
                 };
+                Some(Instruction::Binary(binary))
+            } else {
+                None
+            };
 
-                eval_constant_binary_op(lhs, rhs, binary.operator, operand_type).is_some()
+            if let Some(new_instruction) = new_instruction {
+                self.inserter.function.dfg[instruction_id] = new_instruction;
             }
-            _ => false,
         }
     }
 
+    fn eval_binary_op_with_induction_vars(
+        &self,
+        binary: &Binary,
+        induction_vars: &HashMap<ValueId, FieldElement>,
+    ) -> bool {
+        if !matches!(binary.operator, BinaryOp::Add { .. } | BinaryOp::Mul { .. }) {
+            return false;
+        }
+
+        let operand_type = self.inserter.function.dfg.type_of_value(binary.lhs).unwrap_numeric();
+
+        let lhs_const = self.inserter.function.dfg.get_numeric_constant_with_type(binary.lhs);
+        let rhs_const = self.inserter.function.dfg.get_numeric_constant_with_type(binary.rhs);
+        let (lhs, rhs) = match (
+            lhs_const,
+            rhs_const,
+            induction_vars.get(&binary.lhs),
+            induction_vars.get(&binary.rhs),
+        ) {
+            (Some((lhs, _)), None, None, Some(upper_bound)) => (lhs, *upper_bound),
+            (None, Some((rhs, _)), Some(upper_bound), None) => (*upper_bound, rhs),
+            _ => return false,
+        };
+
+        eval_constant_binary_op(lhs, rhs, binary.operator, operand_type).is_some()
```

Comment:
This could do with some documentation on what this call tells us, i.e. 

> We evaluate this expression using the upper bounds of its inputs to check whether it will ever overflow. If so, this will cause `eval_constant_binary_op` to return `None`. Therefore a `Some` value shows that this operation is safe.

Reply:
done